stats/stats_test.go

```diff
@@ -0,0 +1,1145 @@
+	}
+	if d.ctx == nil {
+		t.Fatalf("d.ctx = nil, want <non-nil>")
+	}
+	// TODO check real length, not just > 0.
+	if st.WireLength <= 0 {
+		t.Fatalf("st.Lenght = 0, want > 0")
+	}
+	if d.client {
+		if st.FullMethod != e.method {
+			t.Fatalf("st.FullMethod = %s, want %v", st.FullMethod, e.method)
+		}
+		if st.RemoteAddr.String() != e.serverAddr {
+			t.Fatalf("st.LocalAddr = %v, want %v", st.LocalAddr, e.serverAddr)
+		}
+		if st.Encryption != e.encryption {
+			t.Fatalf("st.Encryption = %v, want %v", st.Encryption, e.encryption)
+		}
+	}
+}
+
+func checkOutPayload(t *testing.T, d *gotData, e *expectedData) {
+	var (
+		ok bool
+		st *stats.OutPayload
+	)
+	if st, ok = d.s.(*stats.OutPayload); !ok {
+		t.Fatalf("got %T, want OutPayload", d.s)
+	}
+	if d.ctx == nil {
+		t.Fatalf("d.ctx = nil, want <non-nil>")
+	}
+	if d.client {
+		b, err := proto.Marshal(e.requests[e.reqIdx])
+		if err != nil {
+			t.Fatalf("failed to marshal message: %v", err)
+		}
+		if reflect.TypeOf(st.Payload) != reflect.TypeOf(e.requests[e.reqIdx]) {
+			t.Fatalf("st.Payload = %T, want %T", st.Payload, e.requests[e.reqIdx])
+		}
+		e.reqIdx++
+		if string(st.Data) != string(b) {
+			t.Fatalf("st.Data = %v, want %v", st.Data, b)
+		}
+		if st.Length != len(b) {
+			t.Fatalf("st.Lenght = %v, want %v", st.Length, len(b))
+		}
+	} else {
+		b, err := proto.Marshal(e.responses[e.respIdx])
+		if err != nil {
+			t.Fatalf("failed to marshal message: %v", err)
+		}
+		if reflect.TypeOf(st.Payload) != reflect.TypeOf(e.responses[e.respIdx]) {
+			t.Fatalf("st.Payload = %T, want %T", st.Payload, e.responses[e.respIdx])
+		}
+		e.respIdx++
+		if string(st.Data) != string(b) {
+			t.Fatalf("st.Data = %v, want %v", st.Data, b)
+		}
+		if st.Length != len(b) {
+			t.Fatalf("st.Lenght = %v, want %v", st.Length, len(b))
+		}
+	}
+	// TODO check WireLength and ReceivedTime.
+	if st.SentTime.IsZero() {
+		t.Fatalf("st.SentTime = %v, want <non-zero>", st.SentTime)
+	}
+}
+
+func checkOutTrailer(t *testing.T, d *gotData, e *expectedData) {
+	var (
+		ok bool
+		st *stats.OutTrailer
+	)
+	if st, ok = d.s.(*stats.OutTrailer); !ok {
+		t.Fatalf("got %T, want OutTrailer", d.s)
+	}
+	if d.ctx == nil {
+		t.Fatalf("d.ctx = nil, want <non-nil>")
+	}
+	if st.Client {
+		t.Fatalf("st IsClient = true, want false")
+	}
+	// TODO check real length, not just > 0.
+	if st.WireLength <= 0 {
+		t.Fatalf("st.Lenght = 0, want > 0")
+	}
+}
+
+func checkEnd(t *testing.T, d *gotData, e *expectedData) {
+	var (
+		ok bool
+		st *stats.End
+	)
+	if st, ok = d.s.(*stats.End); !ok {
+		t.Fatalf("got %T, want End", d.s)
+	}
+	if d.ctx == nil {
+		t.Fatalf("d.ctx = nil, want <non-nil>")
+	}
+	if st.EndTime.IsZero() {
+		t.Fatalf("st.EndTime = %v, want <non-zero>", st.EndTime)
+	}
+	if grpc.Code(st.Error) != grpc.Code(e.err) || grpc.ErrorDesc(st.Error) != grpc.ErrorDesc(e.err) {
+		t.Fatalf("st.Error = %v, want %v", st.Error, e.err)
+	}
+}
+
+func TestServerStatsUnaryRPC(t *testing.T) {
```

Comment:
Is there a way we can verify the contexts from the stats of a particular rpc are same?